### ISSUE TITLE
KeyError when using Tokenizer tokens_to_events function

### ISSUE THREAD

**Issue by gonzaloarca** [Natooz/MidiTok#22]
## Description
I am trying to use the `MidiTok` library to tokenize a MIDI file and convert the resulting tokens into events, but I am encountering a `KeyError`. When calling the `tokens_to_events` method and passing it the list of integer IDs generated by the `midi_to_tokens` method, I get the following log:

```
Traceback (most recent call last):
  File "..../main.py", line 6, in <module>
    tokens = tokenize_midi(midi_path)
  File "..../tokenizer.py", line 29, in tokenize_midi
    events = tokenizer.tokens_to_events(tokens)
  File "..../.local/share/virtualenvs/midi2spectrogram-dataset-rsI0vtcY/lib/python3.10/site-packages/miditok/midi_tokenizer_base.py", line 196, in tokens_to_events
    name, val = self.vocab[i].token_to_event[token].split('_')
KeyError: 34
```

## Steps to reproduce
Here is the source code I am using:

```python
pitch_range = range(21, 109)
beat_res = {(0, 4): 8, (4, 12): 4}
nb_velocities = 32
additional_tokens = {
    'Chord': True,
    'Tempo': True,
    'Program': True,
    'Rest': False,
    'TimeSignature': False,
    'nb_tempos': 32,  # nb of tempo bins
    'tempo_range': (40, 250)}  # (min, max)

tokenizer = MuMIDI(pitch_range, beat_res, nb_velocities,
                   additional_tokens, mask=True)
midi = MidiFile(midi_path)

tokens = tokenizer.midi_to_tokens(midi)

events = tokenizer.tokens_to_events(tokens)

for event in events:
    print(event)
```

Please let me know if there's any other information I can provide that would be helpful in resolving this issue. The MIDI file used is attached below.

[test.mid.zip](https://github.com/Natooz/MidiTok/files/10407286/test.mid.zip)

**Comment by Natooz** [comment]
Hi @gonzaloarca , thanks for the report !

I just replicated and I located the issue, I'm working on it and will push the fix later today !

**Comment by Natooz** [comment]
Here you go ! 947af8cfa0c72212a8835e10fe1f804356d13e8a

I am currently working on other features, the next version (1.4.0) will be released very soon (today if I'm very productive), so the fix for MuMIDI will be included within.

I must also say that MuMIDI is more complex to use with neural networks than other tokenizations. Not only it required a pooling operation on the embeddings, but the number of embeddings may vary. You can't for instance directly convert the tokens returned by MidiTok to PyTorch / Tensorflow tensors as the list doesn't have a uniform shape. This has to be done "time-step per time-step" which might be slower.

**Comment by gonzaloarca** [comment]
Wow! Thanks for replying and fixing on such short notice, I tremendously appreciate it.

I've been trying to use this library for creating a dataset for fine-tuning HuggingFace's Stable Diffusion model. The idea is to use an [already existing dataset of multitrack songs in the audio domain (from the Million Song Dataset) and their corresponding MIDI transcriptions](https://colinraffel.com/projects/lmd/), and map all the audios to spectrograms and the MIDI transcriptions into strings, acting as the captions for the spectrograms. 

Ideally SD would pickup the MIDI events as features in the spectrograms, and would then generate new spectrograms when fed new MIDI strings. 

I'd really like to hear your opinion on the matter, and please let me know if you have any tips for using this tokenizer to fine-tune HuggingFace's Stable Diffusion.

Have a good one!

**Comment by Natooz** [comment]
I am not sure that SD, or any architecture with classifier (free) guidance or contrastive learning training objective is best suited for this task, as the MIDI -> audio synthesis is "kind of determinist". The task does not imply to generate new or original content. In case you missed them, you might be interest in [DDSP](https://magenta.tensorflow.org/ddsp) and [MIDI-DDSP](https://magenta.tensorflow.org/midi-ddsp). The former is a set of differentiable DSP functions, that significantly improves the performances of audio music synthesis. The latter is a model that achieves MIDI -> audio synthesis (and the other way).

Nevertheless the results with SD could still be interesting and I guess someone has to try to find out !
Adapting the HF SD model for it can probably be achieved, but this would require not just a "fine-tuning" but a whole training as the modalities are completely different.

For the caption side, I would preferably use a "normal" MIDI tokenization (as opposed to what I call embedding pooling strategies such as CPWord or MuMIDI, where you first convert tokens to embeddings, then merge/pool the embeddings, code example here https://github.com/Natooz/MidiTok/issues/13#issuecomment-1126423904). Using embedding pooling techniques would require to override the input layers of the SD, which is probably possible but adds some complexity. I would stick to TSD / REMI / Structured tokenization, with some Byte Pair Encoding (BPE).

For the image / spectrogram, override the output layers if necessary to that the network output shapes matches the spectrograms. (if needed, I did not tried to use / train SD so I don't know really)

**Comment by gonzaloarca** [comment]
That makes perfect sense, thanks a lot! I did check out MIDI-DDSP, but it can only synthesize monophonic tracks, and for a very limited variety of orchestra-type instruments, which is not very useful for a more general purpose synthesizing task.

Are you aware of any other MIDI -> audio synthesis models? I've been using [Multi-instrument Music Synthesis with Spectrogram Diffusion](https://github.com/magenta/music-spectrogram-diffusion), which is probably the most complete model I've found yet, but the synthesized parts sound a little too robotic/fake, and could definitely be outperformed by any deterministic emulating software.
